rt/features/metrics/src/main/java/org/apache/cxf/metrics/micrometer/MicrometerMetricsContext.java

```diff
@@ -0,0 +1,138 @@
+import org.apache.cxf.message.Message;
+import org.apache.cxf.metrics.MetricsContext;
+import org.apache.cxf.metrics.micrometer.provider.TagsCustomizer;
+import org.apache.cxf.metrics.micrometer.provider.TagsProvider;
+import org.apache.cxf.metrics.micrometer.provider.TimedAnnotationProvider;
+
+import io.micrometer.core.annotation.Timed;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Tag;
+import io.micrometer.core.instrument.Timer;
+
+public class MicrometerMetricsContext implements MetricsContext {
+
+    private final MeterRegistry registry;
+    private final TagsProvider tagsProvider;
+    private final TimedAnnotationProvider timedAnnotationProvider;
+    private final List<TagsCustomizer> tagsCustomizers;
+
+    private final String metricName;
+    private final boolean autoTimeRequests;
+
+    public MicrometerMetricsContext(MeterRegistry registry, TagsProvider tagsProvider,
+                                    TimedAnnotationProvider timedAnnotationProvider,
+                                    List<TagsCustomizer> tagsCustomizers, String metricName, boolean autoTimeRequests) {
+        this.registry = registry;
+        this.tagsProvider = tagsProvider;
+        this.timedAnnotationProvider = timedAnnotationProvider;
+        this.tagsCustomizers = tagsCustomizers;
+        this.metricName = metricName;
+        this.autoTimeRequests = autoTimeRequests;
+    }
+
+    @Override
+    public void start(Exchange ex) {
+        Message request = ex.getInMessage();
+        TimingContext timingContext = TimingContext.get(request);
+        if (timingContext == null) {
+            startAndAttachTimingContext(request);
+        }
+    }
+
+    @Override
+    public void stop(long timeInNS, long inSize, long outSize, Exchange ex) {
+        Message request = ex.getInMessage();
+        TimingContext timingContext = TimingContext.get(request);
+        if (timingContext == null) {
+            timingContext = startAndAttachTimingContext(request);
+        }
+        record(timingContext, ex);
+    }
+
+    private TimingContext startAndAttachTimingContext(Message request) {
+        Timer.Sample timerSample = Timer.start(this.registry);
+        TimingContext timingContext = new TimingContext(timerSample);
+        timingContext.attachTo(request);
+        return timingContext;
+    }
+
+    private void record(TimingContext timingContext, Exchange ex) {
+        Set<Timed> annotations = timedAnnotationProvider.getTimedAnnotations(ex);
+        Timer.Sample timerSample = timingContext.getTimerSample();
+        Supplier<Iterable<Tag>> tags = () -> {
+            Iterable<Tag> defaultTags = this.tagsProvider.getTags(ex);
+            List<Tag> additionalTags = new ArrayList<>();
+            tagsCustomizers
+                    .forEach(tagsCustomizer -> tagsCustomizer.getAdditionalTags(ex).forEach(additionalTags::add));
+            return Iterables.concat(defaultTags, additionalTags);
```

Comment:
Bringing whole Guava for using `Iterables` seems too much, could you please rewrite it using standard library? Fe:

```
            return () -> Stream.concat(
                    StreamSupport.stream(defaultTags.spliterator(), false), 
                    StreamSupport.stream(additionalTags.spliterator(), false))
                .iterator();
```

Reply:
Fixed, thank you. (Un)fortunately, I have not been working with Iterables for 6 years. :)